crates/ra_assists/src/handlers/replace_unwrap_with_match.rs

```diff
@@ -0,0 +1,188 @@
+    let name = method_call.name_ref()?;
+    if name.text() != "unwrap" {
+        return None;
+    }
+    let caller = method_call.expr()?;
+    let ty = ctx.sema.type_of_expr(&caller)?;
+
+    let type_name = ty.as_adt()?.name(ctx.sema.db);
+    if type_name == name![Result] {
+        let ok_path = make::path_unqualified(make::path_segment(make::name_ref("Ok")));
+        let bind_pat = make::bind_pat(make::name("a")).into();
+        let ok_tuple = make::tuple_struct_pat(ok_path, iter::once(bind_pat)).into();
+        let bind_path = make::path_unqualified(make::path_segment(make::name_ref("a")));
+
+        let ok_arm = make::match_arm(iter::once(ok_tuple), make::expr_path(bind_path));
+
+        let unreachable_call = make::unreachable_macro_call().into();
+        let err_arm = make::match_arm(iter::once(make::placeholder_pat().into()), unreachable_call);
+        let match_arm_list = make::match_arm_list(vec![ok_arm, err_arm]);
+        let match_expr = make::expr_match(method_call.expr()?, match_arm_list);
+        let match_expr = IndentLevel::from_node(method_call.syntax()).increase_indent(match_expr);
+
+        return ctx.add_assist(AssistId("replace_unwrap_with_match"), "Replace unwrap with match", |edit| {
+            edit.target(method_call.syntax().text_range());
+            edit.set_cursor(caller.syntax().text_range().start());
+            edit.replace_ast::<ast::Expr>(method_call.into(), match_expr);
+        });
+    }
+
+    if type_name == name![Option] {
+        let some_path = make::path_unqualified(make::path_segment(make::name_ref("Some")));
+        let bind_pat = make::bind_pat(make::name("a")).into();
+        let some_tuple = make::tuple_struct_pat(some_path, iter::once(bind_pat)).into();
+        let bind_path = make::path_unqualified(make::path_segment(make::name_ref("a")));
+        let some_arm = make::match_arm(iter::once(some_tuple), make::expr_path(bind_path));
+
+        let unreachable_call = make::unreachable_macro_call().into();
+        let none_arm = make::match_arm(iter::once(make::placeholder_pat().into()), unreachable_call);
+        let match_arm_list = make::match_arm_list(vec![some_arm, none_arm]);
+        let match_expr = make::expr_match(method_call.expr()?, match_arm_list);
+        let match_expr = IndentLevel::from_node(method_call.syntax()).increase_indent(match_expr);
+
+        return ctx.add_assist(AssistId("replace_unwrap_with_match"), "Replace unwrap with match", |edit| {
+            edit.target(method_call.syntax().text_range());
+            edit.set_cursor(caller.syntax().text_range().start());
+            edit.replace_ast::<ast::Expr>(method_call.into(), match_expr);
+        });
+    }
+
+    None
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::helpers::{check_assist, check_assist_target};
+
+    #[test]
+    fn test_replace_result_unwrap_with_match() {
+        check_assist(
+            replace_unwrap_with_match,
+            "
```

Comment:
Note that if you use a `r"` literal here, you'll get syntax highlighting (provided that semanticHighlighitng is enabled in settings). 

Perhaps we should enabled this for all literals, and not only for `r"`. Or perhaps we should proacively normalize all rust tests to use raw strings... I am unsure about this at the moment.